### PR TITLE
Get end points

### PR DESCRIPTION
**Pull request by RaphealOlumofe**
Retreive a single car by sending ID
Batch retrieve cars by List<ID>

### CODE REVIEW

**Review by petermcneil — CHANGES_REQUESTED** [review]
Every change but the one to add a GET endpoint should be removed. If it's unclear what that is please have a chat with me

**Review by petermcneil — CHANGES_REQUESTED** [review]
It also needs unit tests for the GET

File: build.gradle [hunk 1]
```diff
@@ -9,15 +9,16 @@ subprojects {
...
         testImplementation("org.junit.jupiter:junit-jupiter-api:$jupiter")
         testRuntime("org.junit.jupiter:junit-jupiter-engine:$jupiterEngine")
         compile group: 'org.apache.logging.log4j', name: 'log4j-api', version: "$log4j"
         compile group: 'org.apache.logging.log4j', name: 'log4j-core', version: "$log4j"
 
+
     }
+}
+
+dependencies {
```
Comment (timlarochebbc) [inline build.gradle]: Unsure what this extra dependencies is doing, should not be needed

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                returnedCar = car;
+                return returnedCar;
+            }
+            }
+          return null;
+    }
+
+    @GetMapping(path = "/cars", consumes = {"application/json"})
+    @ResponseBody
+    private List<Car> getBatchCars(@RequestBody List<Integer> list) {
```
Comment (petermcneil) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java]: This shouldn't be private, did you test this through postman?

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.springframework.stereotype.Component;
+import org.springframework.web.bind.annotation.*;
+import starter.project.domain.Car;
+
+import java.util.*;
+
+@RestController
+@Component
+public class CarController {
```
Comment (petermcneil) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java:13]: where are the other methods? the PUT ones?

File: domain/build.gradle [hunk 4]
```diff
@@ -4,6 +4,14 @@ dependencies {
 
     compile group: 'com.fasterxml.jackson.core', name: 'jackson-core', version: '2.10.3'
     compile group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.10.3'
+    compile 'org.springframework.boot:spring-boot-starter-web:2.0.1.RELEASE'
+    compile 'org.springframework.boot:spring-boot-starter-data-jpa:2.0.1.RELEASE'
+   //compile group: 'com.h2database', name: 'h2', version: '1.3.148'
```
Comment (timlarochebbc) [inline domain/build.gradle]: It's good practice to delete comments like this as they are effectively void and not used, they do not belong in the checked in code as they are bloat. If you want to keep them for yourself to use later I would make a note somewhere

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+
+    private List<Car> carNewList = new ArrayList<>();
+
+    @GetMapping(path ="/car/{carId}", consumes = {"application/json"})
+    @ResponseBody
+    public Car getCars(@PathVariable("carId") int carId) {
+
+        for (Car car : carNewList) {
+            if (car.getId() == (carId)) {
```
Comment (petermcneil) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java:28]: If the ID is stored at the Car level, then you should think about adding a better way to access the data. Perhaps a Map<CarId, Car> would be more suitable.

File: domain/src/test/java/starterproject/domain/TransmissionTest.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package starterproject.domain;
+
+class TransmissionTest {
```
Comment (timlarochebbc) [inline domain/src/test/java/starterproject/domain/TransmissionTest.java]: Remove this class, no need for it at the moment and it doesn't define a useful framework for us (or another programmer) to use later.

 In an ideal world, every class should be tested so, with that in mind, just defining that we need a test is kind of redundant. 

If there was some non-trivial functionality or non-trivial setup ready for a test then maybe we have a case for having it.

File: frontend/build.gradle [hunk 7]
```diff
@@ -16,12 +16,17 @@ repositories {
...
     testCompile group: 'junit', name: 'junit', version: '4.12'
-    implementation 'org.springframework.boot:spring-boot-starter-web'
+    //implementation 'org.springframework.boot:spring-boot-starter-web'
     compile project(':domain')
     testImplementation('org.springframework.boot:spring-boot-starter-test') {
         exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
         implementation 'com.google.code.gson:gson:2.8.6'
-
+        compile 'org.springframework.boot:spring-boot-starter-web:2.0.1.RELEASE'
+//        compile 'org.springframework.boot:spring-boot-starter-data-jpa:2.0.1.RELEASE'
```
Comment (timlarochebbc) [inline frontend/build.gradle]: Same comment about deleting unused comments / comments that effectively do the same as deleting code

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    for(int i= 0; i <= list.size() -1; i ++){
+        if(carNewList.get(i).getId() == list.get(i)){
+            Car car = carNewList.get(i);
+            newList.add(car);
+        }
+    }
+        return newList;
+
+    }
+
```
Comment (timlarochebbc) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java:51]: Empty white space, please make an effort to remove this it might seem pedantic but it's happened a lot of PRs now and it does make it harder to review.

Notice how it makes it look like this file has added a lot (+57 lines) but a lot of them are whitespace.

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            }
+          return null;
+    }
+
+    @GetMapping(path = "/cars", consumes = {"application/json"})
+    @ResponseBody
+    private List<Car> getBatchCars(@RequestBody List<Integer> list) {
+        List<Car> newList = new ArrayList<>();
+    for(int i= 0; i <= list.size() -1; i ++){
+        if(carNewList.get(i).getId() == list.get(i)){
```
Comment (timlarochebbc) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java]: This logic will only work if the cars in `carNewList` are defined in the same order as the list of the request body. Let me know if you have trouble spotting the bug but think about what the value `i` is in relation to the two lists.

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    @ResponseBody
+    public Car getCars(@PathVariable("carId") int carId) {
+
+        for (Car car : carNewList) {
+            if (car.getId() == (carId)) {
+                Car returnedCar;
+                returnedCar = car;
+                return returnedCar;
+            }
+            }
```
Comment (timlarochebbc) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java:32]: Minor comment, line up your style braces for readability:
```
for() {
   if() {
   }
}
```

**Review by petermcneil — CHANGES_REQUESTED** [review]
There is no way to test this due to the fact that the PUT endpoints have disappeared.

However I tried hitting the endpoints with some IDs anyway.

`/car/{id}` endpoint returns 200 even though it doesn't return anything. This should be a 404 as the Car is not found in the list.

`/cars` is private, when it should be public. It also takes a body, this is not how REST APIs are specced up. The body should be ignored in this case. Therefore you need to pass the ids in as a comma delimited list in a query parameter

A stack overflow question to show what I mean.
https://stackoverflow.com/a/9372278

https://stackoverflow.com/questions/978061/http-get-with-request-body

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java (hunk 5 above)
Comment (timlarochebbc) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java:28]: This logic is fine, the only thing is I don't think carNewList will ever have objects in it? So functionally the logic is fine but we will never get a car back. Perhaps give carNewList some initial values so we can see it working.

File: frontend/src/test/java/olumofe/raphael/frontend/CarControllerTest.java [hunk 11]
```diff
@@ -29,6 +28,8 @@ public void checkObjectIsCreated() throws Exception {
 
         ResponseEntity<String> response = restTemplate.getForEntity(
                 new URL("http://localhost:" + port + "/cars").toString(), String.class);
+        // assertEquals("Hello Controller", response.getBody());
         assertNotNull(response.getBody());
```
Comment (timlarochebbc) [inline frontend/src/test/java/olumofe/raphael/frontend/CarControllerTest.java]: This is not a useful assertion. We should check if the input we give produces the output we expect. This is testing whether we get something at all (which is sometimes useful).

**Review by timlarochebbc — COMMENTED** [review]
Left some comments. Please try to get rid of excess whitespace for readability. Also, what happened to your PUT endpoints? This work should be on top of that.

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java (hunk 3 above)
Comment (RaphealOlumofe) [reply frontend/src/main/java/olumofe/raphael/frontend/CarController.java:13]: After we resolved the conflicts yesterday they disappeared.

Comment (timlarochebbc) [reply frontend/src/main/java/olumofe/raphael/frontend/CarController.java:13]: You should be able to pick and choose what to keep and delete when you resolve conflicts. What did you merge?

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java (hunk 5 above)
Comment (RaphealOlumofe) [reply frontend/src/main/java/olumofe/raphael/frontend/CarController.java:28]: Ive  now created a Post method, and ive  it tested it to ensure that it works

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java (hunk 9 above)
Comment (RaphealOlumofe) [reply frontend/src/main/java/olumofe/raphael/frontend/CarController.java]: The methods been changed, thanks

**Review by petermcneil — CHANGES_REQUESTED** [review]
- Why is there a double loop in the getBatchCar method? This should be a direct access based on the ID, same goes for getCars
- Why are there @ResponseBody everywhere?
- There is a @Component and @RestController on the CarController class, choose the one that’s needed as I think RestController already contains a Component
- The HashMap in createCarList has types specified in the new HashMap<..., this has been obselete for 10+ years, remove them
- For some reason the package name has changed, on master it is starter.project.frontend
- You need to add unit tests for this, covering the scenarios that were laid out in all the tickets.
- Semantically the names of the methods aren’t clear. car() method doesn’t really tell anyone what it does.
-  Un clear why the variable assignment here is split up.
Car returnedCar;
                returnedCar = car;
- There needs to be error handling for when a car can’t be found for the ID. If a car can’t be found return 404. Similar to how websites would display it if there were nothing to be found, https://github.com/404
- There is still somehow `.idea` folder being included

**Comment by timlarochebbc** [comment]
First off, everything is starting to look a lot better so far, good job! Quick note (I'm still having a look at everything), can you see how all your commit messages are "`get endpoints`". This is really hard to follow your work and what you've done. 

Think about your commit messages being like a log or journal of your work work. When you commit you are writing a message in the log about what you've done since the last time you've committed so that you can read the log in some chronological order and see what's been done overall.

When you merge your PR, your commit messages get combined into the master log so having 20 non-useful messages is going to make it harder to find an important change.

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+                Car returnedCar = car;
+                return returnedCar;
+            }
+        }
+        return null;
+    }
+    @GetMapping(path = "/cars/{list}")
+    public List<Car> getBatchCars(@PathVariable List<Integer> list) {
+        List<Car> newList = new ArrayList<>();
+        for (int i = 0; i < carNewList.size(); i++)
```
Comment (timlarochebbc) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java]: Great! This algorithm works but it's not the most efficient... is there a way we can utilise a data structure to speed this up?

 To get slightly nerdy this is `O(n^2)` run time which does not scale very well!

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @PostMapping(path = "/cars", consumes = {"application/json"})
+    public Map<String, Car> createCarList(@RequestBody List<Car> carList) {
+        carNewList = carList;
+        Map<String, Car> map = new HashMap<>();
+        for (int i = 0; i < carList.size(); i++) {
+            map.put(String.valueOf(i), carNewList.get(i));
+        }
+        return map;
```
Comment (timlarochebbc) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java:55]: This is not bad, but we only utilise the map inside this method, perhaps we can think about using it throughout the whole class. Maybe instead of another data structure to speed things up?

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    @PostMapping(path = "/car", consumes = {"application/json"})
+    public List<Car> addCar(@RequestBody Car car) {
+        carNewList.add(car);
+        return carNewList;
+    }
+    @GetMapping(path = "/car/{carId}")
+    public Car getCars(@PathVariable("carId") int carId) {
+        for (Car car : carNewList) {
+            if (car.getId() == (carId)) {
+                Car returnedCar = car;
```
Comment (timlarochebbc) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java]: We can simplify this to one line without using a variable!
Don't always listen to intelliJ! :P

**Review by timlarochebbc — CHANGES_REQUESTED** [review]
Left some comments. Deliberately left some of the comments a bit vague as I think working things out helps learning a little bit but of course I'm happy to help, don't feel bad if you can't work it out as we were all there once. Give it a good try first though! 

`.idea` folder is still being included. Lets fix this it should be fast. 
Some whitespace is still hanging on the bottom but overall better, think about whitespace as 
a positive as well, it can be useful in between methods to visually separate their functionality. I think we've gone too far in one direction and it's looking a bit messy everything tightly pushed together. 

```
@Something
func(a)
@Something@SomethingElse
func(b)
func(c)
```
vs 
```
@Something
func(a){
    ...
}

@Something @SomethingElse
func(b){
   ...
}
```

What would be good is some unit tests and some error messages. At the moment your application (probably) doesn't fail but returns an empty screen for bad input etc. We can make this more useful by sending back a 404 in the same way that we send a 200 if we are successful.

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java (hunk 13 above)
Comment (RaphealOlumofe) [reply frontend/src/main/java/olumofe/raphael/frontend/CarController.java:55]: methods been changed

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java (hunk 12 above)
Comment (RaphealOlumofe) [reply frontend/src/main/java/olumofe/raphael/frontend/CarController.java]: method has been changed

File: domain/build.gradle [hunk 15]
```diff
@@ -1,15 +1,12 @@
 
 dependencies {
     annotationProcessor 'org.projectlombok:lombok:1.18.12'
-
     compile group: 'com.fasterxml.jackson.core', name: 'jackson-core', version: '2.10.3'
     compile group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.10.3'
-
-
-
+    compile 'org.springframework.boot:spring-boot-starter-web:2.0.1.RELEASE'
```
Comment (ejalmeida) [inline domain/build.gradle:6]: We should be consistent when adding new dependencies and follow the `group`-`name`-`version` declaration style

File: domain/src/main/java/starter/project/domain/Car.java [hunk 16]
```diff
@@ -5,12 +5,18 @@
 import lombok.Data;
 import lombok.NoArgsConstructor;
 
+import javax.persistence.Entity;
+import javax.persistence.Id;
```
Comment (ejalmeida) [inline domain/src/main/java/starter/project/domain/Car.java:9]: These two imports are not being used. You can have intellij optimise the imports for this class

File: domain/src/main/java/starter/project/domain/Doors.java [hunk 17]
```diff
@@ -6,11 +6,14 @@
 import lombok.Data;
 import lombok.NoArgsConstructor;
 
+import javax.persistence.Embeddable;
```
Comment (ejalmeida) [inline domain/src/main/java/starter/project/domain/Doors.java:9]: This import is not being used, it can be removed

File: domain/src/main/java/starter/project/domain/Transmission.java [hunk 18]
```diff
@@ -6,11 +6,14 @@
 import lombok.Data;
 import lombok.NoArgsConstructor;
 
+import javax.persistence.Embeddable;
```
Comment (ejalmeida) [inline domain/src/main/java/starter/project/domain/Transmission.java:9]: Not used

File: domain/src/main/java/starter/project/domain/Wheels.java [hunk 19]
```diff
@@ -4,11 +4,15 @@
 import lombok.Builder;
 import lombok.Data;
 import lombok.NoArgsConstructor;
+
+import javax.persistence.Embeddable;
```
Comment (ejalmeida) [inline domain/src/main/java/starter/project/domain/Wheels.java:8]: Not used

File: domain/build.gradle [hunk 20]
```diff
@@ -1,15 +1,12 @@
...
+    compile 'org.springframework.boot:spring-boot-starter-data-jpa:2.0.1.RELEASE'
```
Comment (ejalmeida) [inline domain/build.gradle:7]: We probably do not need these dependencies at this stage since we are still using in-memory data structures to hold the data.

File: frontend/build.gradle [hunk 21]
```diff
@@ -2,26 +2,21 @@ plugins {
...
     testCompile group: 'junit', name: 'junit', version: '4.12'
-    implementation 'org.springframework.boot:spring-boot-starter-web'
     compile project(':domain')
     testImplementation('org.springframework.boot:spring-boot-starter-test') {
         exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
         implementation 'com.google.code.gson:gson:2.8.6'
-
+        compile 'org.springframework.boot:spring-boot-starter-web:2.0.1.RELEASE'
+       compile group: 'com.h2database', name: 'h2', version: '1.3.148'
+        compile group: 'org.hibernate.javax.persistence', name: 'hibernate-jpa-2.1-api', version: '1.0.0.Final'
```
Comment (ejalmeida) [inline frontend/build.gradle:19]: We probably do not need these dependencies (h2 and hibernate) at this stage since we are still using in-memory data structures to hold the data. We can add them later when we add the database service.

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+    public Car getCar(@PathVariable("carId") int carId) {
+
+        for (Car car : carNewList) {
+            if (car.getId() == (carId)) {
+                return car;
+            } else {throw new ResponseStatusException(
+                    HttpStatus.NOT_FOUND, "entity not found");
+            }
+        }
+        return null;
```
Comment (ejalmeida) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java:34]: We should not return a `null` in this case, probably a 404. I would remove thois `else` branch:
```java
else {throw new ResponseStatusException(
                    HttpStatus.NOT_FOUND, "entity not found");
            }
```
And throw the exception at this point inconditionally

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @GetMapping(path = "/cars/{list}")
+    public List<Car> getBatchCars(@PathVariable List list) {
```
Comment (ejalmeida) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java:38]: I did not know you could add a list as a path variable, quite cool. It is fine as long as it is working.
So how would requests look like? `/cars/list/1,2,3`

File: domain/src/main/java/starter/project/domain/Car.java (hunk 16 above)
Comment (RaphealOlumofe) [reply domain/src/main/java/starter/project/domain/Car.java:9]: resolved

File: domain/src/main/java/starter/project/domain/Transmission.java (hunk 18 above)
Comment (RaphealOlumofe) [reply domain/src/main/java/starter/project/domain/Transmission.java:9]: resolved

File: domain/build.gradle (hunk 20 above)
Comment (RaphealOlumofe) [reply domain/build.gradle:7]: resolved

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+        return newList;
+    }
+    @PostMapping(path = "/cars", consumes = {"application/json"})
+    public Map<String, Car> createCarList(@RequestBody List<Car> carList) {
+        carNewList = carList;
+
+        for (int i = 0; i < carList.size(); i++) {
+            map.put(String.valueOf(i), carNewList.get(i));
```
Comment (ejalmeida) [inline frontend/src/main/java/olumofe/raphael/frontend/CarController.java:53]: Now the class `Car` has and `id` field, so we should use it as the key to index cars in the map.

File: domain/src/main/java/starter/project/domain/Doors.java (hunk 17 above)
Comment (RaphealOlumofe) [reply domain/src/main/java/starter/project/domain/Doors.java:9]: resolved

File: domain/src/main/java/starter/project/domain/Wheels.java (hunk 19 above)
Comment (RaphealOlumofe) [reply domain/src/main/java/starter/project/domain/Wheels.java:8]: resolved

**Review by ejalmeida — CHANGES_REQUESTED** [review]
* A few comments can be found below

File: frontend/build.gradle (hunk 21 above)
Comment (RaphealOlumofe) [reply frontend/build.gradle:19]: resolved

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java (hunk 22 above)
Comment (RaphealOlumofe) [reply frontend/src/main/java/olumofe/raphael/frontend/CarController.java:34]: resolved

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java (hunk 23 above)
Comment (RaphealOlumofe) [reply frontend/src/main/java/olumofe/raphael/frontend/CarController.java:38]: /cars/1,2.. etc

File: frontend/src/main/java/olumofe/raphael/frontend/CarController.java (hunk 24 above)
Comment (RaphealOlumofe) [reply frontend/src/main/java/olumofe/raphael/frontend/CarController.java:53]: Im going to remove the id field from the Car class
